### PR TITLE
Indices + error handling

### PR DESCRIPTION
**Pull request by daniel-sc**

### CODE REVIEW

File: src/Flow/Mongo/MongoFile.php [hunk 1]
```diff
@@ -90,12 +90,22 @@ public function saveChunk()
...
-            $chunk['data'] = new \MongoBinData($data);
+            $chunk['data'] = new \MongoBinData($data, 0); // \MongoBinData::GENERIC is not defined for older mongo drivers
             $this->config->getGridFs()->chunks->findAndModify($chunkQuery, $chunk, [], ['upsert' => true]);
             unlink($file['tmp_name']);
 
+            $this->ensureIndices();
+
             return true;
         } catch (\Exception $e) {
+            error_log("Could not store chunk: " . $e->getMessage() . "\n" . $e->getTraceAsString());
```
Comment (AidasK) [inline src/Flow/Mongo/MongoFile.php]: do we need an error log here?

File: src/Flow/Mongo/MongoFile.php [hunk 2]
```diff
@@ -90,12 +90,22 @@ public function saveChunk()
...
+            try {
+                if (isset($chunkQuery)) {
+                    $this->config->getGridFs()->chunks->remove($chunkQuery);
+                }
+            } catch (\Exception $e2) {
```
Comment (AidasK) [inline src/Flow/Mongo/MongoFile.php]: Could we catch a specific error and not base exception?

File: src/Flow/Mongo/MongoFile.php (hunk 1 above)
Comment (daniel-sc) [inline src/Flow/Mongo/MongoFile.php]: To me it is meaningful, as something unexpected happened (db down, ...). If this violates coding rules for this project, I can of course remove it.

File: src/Flow/Mongo/MongoFile.php (hunk 2 above)
Comment (daniel-sc) [inline src/Flow/Mongo/MongoFile.php]: Why I catch the base exception here: Whatever went wrong inserting the chunk might as well fail when we try to remove the chunk - hence we catch the same exception to assure the method signature (returning true/false, but no known exceptions). Does that make sense to you?

File: src/Flow/Mongo/MongoFile.php (hunk 1 above)
Comment (AidasK) [inline src/Flow/Mongo/MongoFile.php]: But in most cases we use framework specific error logging functions.

**Comment by AidasK** [comment]
Could you also add mongo code base to phpunit ignore list? Because it has not unit tests. Thanks

File: src/Flow/Mongo/MongoFile.php (hunk 1 above)
Comment (daniel-sc) [inline src/Flow/Mongo/MongoFile.php]: hm - not sure I understand what you mean - in this method there is no access to any framework specific logging. This was meant to be similar to `\move_uploaded_file`, that issues a warning. Would it be better to user `trigger_error("Could not ...", E_USER_WARNING)`?

**Comment by daniel-sc** [comment]
Hi @AidasK 
do the changes resolve the remarks you made?

**Comment by AidasK** [comment]
I still don't like error_log in the code, and this implementation is still not stable, but I can merge it.

**Comment by daniel-sc** [comment]
Hi @AidasK 
I removed the error_log in favour of a E_USER_WARNING

**Comment by AidasK** [comment]
In most php frameworks triggered warnings are just converted to exceptions.
I think, best solution in this case would be just to rethrow the exception.
Catch block:
```php
....
} catch (\Exception $e) {
                if (isset($chunkQuery)) {
                    $this->config->getGridFs()->chunks->remove($chunkQuery);
                }
                throw $e;
}
...

```

**Comment by daniel-sc** [comment]
Hi @AidasK 
if we throw that exception, shouldn't we declare that (in phpdoc) in https://github.com/flowjs/flow-php-server/blob/master/src/Flow/File.php#L107 ? (And as well add some try/catch to https://github.com/flowjs/flow-php-server/blob/master/src/Flow/Basic.php#L40 ?)
(don't want to be picky - just trying to understand and learn..)

**Comment by AidasK** [comment]
No, we don't need phpdoc there, because exception can be thrown only by Mongo extension.

And we don't need to catch exception anywhere,  because application should fail and crash if something unexpected happened. Error will be logged by application logger.

If you want, you can catch this exception in your own application and do some logging.

**Comment by AidasK** [comment]
Thanks!

**Pull request merged**